src/Toml/Codec/Combinators.hs

```diff
@@ -517,33 +518,86 @@ map keyCodec valCodec key = Codec input output
         dict <$ modify updateAction
 
 -- TODO: add docs and tests
-tableMapCodec
-    :: forall key val . Ord key
-    => TomlBiMap Key key  -- ^ Bidirectional converter between TOML and Map keys
-    -> TomlBiMap val AnyValue  -- ^ Bidirectional converter between TOML and Map values
-    -> Key  -- ^ Table name key
-    -> TomlCodec (Map key val)
+tableMapCodec :: forall key val .
+                 Ord key
+              => TomlBiMap Key key  -- ^ Bidirectional converter between TOML and Map keys
+              -> TomlBiMap val AnyValue  -- ^ Bidirectional converter between TOML and Map values
+              -> Key  -- ^ Table name key
+              -> TomlCodec (Map key val)
 tableMapCodec keyBiMap valBiMap key = Codec input output
   where
     input :: TomlEnv (Map key val)
     input = do
-        mTable <- asks $ Prefix.lookup key . tomlTables
-        case mTable of
-            Nothing   -> pure Map.empty
-            Just toml -> fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(tomlKey, anyVal) ->
-                case backward valBiMap anyVal of
-                    Right v  -> case forward keyBiMap tomlKey of
-                        Right k  -> pure (k, v)
-                        Left err -> throwError $ BiMapError err
-                    Left err -> throwError $ BiMapError err
+      mTable <- asks $ Prefix.lookup key . tomlTables
+      case mTable of
+        Nothing   -> pure Map.empty
+        Just toml -> fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(tomlKey, anyVal) ->
+          case backward valBiMap anyVal of
+            Right v  -> case forward keyBiMap tomlKey of
+              Right k  -> pure (k, v)
+              Left err -> throwError $ BiMapError err
+            Left err -> throwError $ BiMapError err
+
+    output :: Map key val -> TomlState (Map key val)
+    output dict = do
+      newToml <- foldM update mempty (Map.toList dict)
+      dict <$ modify (insertTable key newToml)
+
+    update :: TOML -> (key, val) -> TomlState TOML
+    update toml (k, v) = do
+      tomlKey <- MaybeT $ pure $ either (const Nothing) Just $ backward keyBiMap k
+      anyVal <- MaybeT $ pure $ either (const Nothing) Just $ forward valBiMap v
+      return $ insertKeyAnyVal tomlKey anyVal toml
+
+dynamicMapCodec :: forall key val .
+                   Ord key
+                => TomlBiMap Key key
+                -> (Key -> TomlCodec val)
+                -> Key
+                -> TomlCodec (Map key val)
```

Comment:
According to our style guide, the indentation should not depend on the length of the identifier, so this type signature should be formatted differently:

```haskell
dynamicMapCodec
    :: forall
    .  Ord
    => ...
    -> ...
```